src/headers/tfm.h

```diff
@@ -21,8 +21,53 @@
  * Patch
  * Development - 00=release, 01=in-development
  */
-#define TFM_VERSION     0x000D0101
-#define TFM_VERSION_S   "v0.13.1-next"
+#define TFM_VERSION_MAJ    0
+#define TFM_VERSION_MIN    13
+#define TFM_VERSION_PAT    1
+#define TFM_VERSION_DEV    1
+
+
+#define TFM_VERSION     PRIVATE__TFM_VERSION_4(TFM_VERSION_MAJ, \
+                                               TFM_VERSION_MIN, \
+                                               TFM_VERSION_PAT, \
+                                               TFM_VERSION_DEV)
+
+#define TFM_VERSION_S   PRIVATE__TFM_CONC(TFM_VERSION_MAJ, \
+                                          TFM_VERSION_MIN, \
+                                          TFM_VERSION_PAT, \
+                                          TFM_VERSION_DEV)
+
+/* Please use the `TFM_VERSION_3()` macro if you want to compile-time check
+ * for a specific TFM version.
+ * Your code could look as follows:
+
+#if TFM_VERSION <= TFM_VERSION_3(0, 13, 1)
+// do stuff to work with old TFM
+#else
+// do stuff to work with new TFM
+#endif
+
+ */
+
+#define TFM_VERSION_3(maj, min, pat) PRIVATE__TFM_VERSION_4(maj, min, pat, 0xff)
```

Comment:
Hmmm...  this doesn't work so well if I want to check for older than a target release, like this:

``` C
#if TFM_VERSION < TFM_VERSION_3(0,14,0)
/* do stuff to work with old TFM */
#else
/* do stuff to work with fresh TFM */
#endif
```

So the requirement to use this is currently that you must know the latest old versionm rather than knowing the current target version...  and it seems to me that the latest old version is a bit more a moving target, at least potentially.

Reply:
Ah well, I can live with that, come to think of it.  Just gotta have an absurd number in the third position

Reply:
To have that kind of feedback is exactly why I've opened a PR instead of simply committing to develop :)

> Hmmm... this doesn't work so well if I want to check for older than a target release

hmm, so you want to check in your code for a potential next release that you don't know whether it will exist? Shouldn't we usually write code against what we know? That's a serious question, especially since there should be no "absurd numbers" in any position IMO :)

It's very well possible that I didn't think enough while writing this and `TFM_VERSION_3()` should behave differently in order to be usable. Maybe it would also make more sense to have a `TFM_VERSION_AT_LEAST(min, maj, pat)` which does the comparison operation?

Reply:
> hmm, so you want to check in your code for a potential next release that you don't know whether it will exist?

Not necessarily.  I want to be able to check against a specific release, already existing or future, doesn't really matter.  The current construction forces me to check for a version *before* my target release, and I end up doing a construction like this, 'cause I can't possibly know that 0.13.1 will be the last before what I target):

``` C
#elif TFM_VERSION <= TFM_VERSION_3(0, 13, 89)
```

After all, some time in the future, I assume that my target release will exist, while I can't possibly know if there's going to be another release before that one.

Reply:
The construction I ended up doing is workable, just a bit surprising compared to other similar macros I've used before (for example `KERNEL_VERSION`)

Reply:
The thing that throws everything off is 0xff, I might add.  Not sure if 0 or 1 is the right value to use here, though...

Reply:
Ah!  What turned out to be the big surprise was that this:

``` C
#if TFM_VERSION < TFM_VERSION_3(0,13,1)
```

Turned out to be true!  That came as a surprise...

Reply:
I've slept on this, and think that the following would be the most sane:

```suggestion
#define TFM_VERSION_3(maj, min, pat) PRIVATE__TFM_VERSION_4(maj, min, pat, 0)
```

The logic is that any preprocessor expression should compare against the *release* version.  Furthermore, the TFM version is really this:

{TFM_VERSION_MAJ}.{TFM_VERSION_MIN}.{TFM_VERSION_PAT} + {TFM_VERSION_DEV}

... where the addition of {TFM_VERSION_DEV} really means that any of the TFM_VERSION components will be incremented.

Therefore, when building an external project against a TFM *release* build, the expression `TFM_VERSION {op} TFM_VERSION_3(TFM_VERSION_MAJ,TFM_VERSION_MIN,TFM_VERSION_PAT)` should result in this:

|  \<   | \=\=  |  \>   |
| ----- | ----- | ----- |
| false | true  | false |

Likewise, when building an external project against a TFM *development* build, the same expression should result in this:

|  \<   | \=\=  |  \>   |
| ----- | ----- | ----- |
| false | false | true  |

Reply:
Does that testcase even make sense? :D

Reply:
It does